### PR TITLE
🫘 Return error from useCall with invalid arguments

### PR DESCRIPTION
**Pull request by jakvbs**
Closes https://github.com/TrueFiEng/useDApp/issues/987

### CODE REVIEW

File: packages/core/src/hooks/useCall.test.tsx [hunk 1]
```diff
@@ -420,6 +421,71 @@ describe('useCall', () => {
...
+                args: [123],
+              }),
+            {
+              config,
+            }
+          )
+          await waitForCurrent((val) => val !== undefined)
+
+          expect(result.current?.value).to.be.undefined
+          expect(result.current?.error).to.exist
```
Comment (rzadp) [inline packages/core/src/hooks/useCall.test.tsx:450]: Check the exact error string

File: packages/core/src/helpers/calls.ts [hunk 2]
```diff
@@ -17,6 +17,17 @@ export function warnOnInvalidCall(call: Call | Falsy) {
   console.warn(`Invalid contract call: address=${contract.address} method=${method} args=${args}`)
 }
 
+function getInvalidRawCall(call: Call, chainId: number): RawCall {
+  const { contract, method, args } = call
+
+  return {
+    address: contract.address,
+    data: `Invalid contract call: address=${contract.address} method=${method} args=${args}`,
```
Comment (rzadp) [inline packages/core/src/helpers/calls.ts:25]: Consider doing `JSON.stringify(args)`

File: packages/core/src/helpers/calls.ts [hunk 3]
```diff
@@ -20,14 +20,41 @@ export function warnOnInvalidCall(call: Call | Falsy) {
 /**
  * @internal Intended for internal use - use it on your own risk
  */
-export function encodeCallData(call: Call | Falsy, chainId: number, queryParams: QueryParams = {}): RawCall | Falsy {
+export interface InvalidRawCall {
+  chainId: ChainId
+  address: string
+  data: string
+  isValid: boolean
```
Comment (rzadp) [inline packages/core/src/helpers/calls.ts:27]: This field does not make sense - there is no `InvalidRawCall` with `isValid=true`

Comment (rzadp) [reply packages/core/src/helpers/calls.ts:27]: How about - `InvalidRawCall extends RawCall`
with an additional field `errorString`

File: packages/core/src/hooks/useRawCalls.ts [hunk 4]
```diff
@@ -15,11 +16,11 @@ import { utils } from 'ethers'
  * @param calls List of calls, also see {@link RawCall}. Calls need to be in the same order across component renders.
  * @returns list of multicall calls. See {@link RawCallResult} and {@link useRawCall}.
  */
-export function useRawCalls(calls: (RawCall | Falsy)[]): RawCallResult[] {
+export function useRawCalls(calls: (RawCall | InvalidRawCall | Falsy)[]): RawCallResult[] {
   const { dispatchCalls, chains } = useContext(MultiChainStatesContext)
 
   useEffect(() => {
-    const filteredCalls = calls.filter(Boolean) as RawCall[]
+    const filteredCalls = calls.filter((call) => !!call && (call as InvalidRawCall).isValid !== false) as RawCall[]
```
Comment (rzadp) [inline packages/core/src/hooks/useRawCalls.ts:23]:
```suggestion
    const filteredCalls = calls.filter((call) => !!call && !(errorString in call)) as RawCall[]
```

File: packages/core/src/helpers/calls.ts [hunk 5]
```diff
@@ -20,29 +20,37 @@ export function warnOnInvalidCall(call: Call | Falsy) {
 /**
  * @internal Intended for internal use - use it on your own risk
  */
-export function encodeCallData(call: Call | Falsy, chainId: number, queryParams: QueryParams = {}): RawCall | Falsy {
+export function validateCall(call: Call | Falsy) {
   if (!call) {
-    return undefined
+    throw new Error('Call is falsy')
```
Comment (rzadp) [inline packages/core/src/helpers/calls.ts:25]: Make sure the behaviour is unchanged.

Add a test for something like:
```
const result = useCall(undefined)
expect(result).to.xxx
```

And make sure the behaviour is the same before and after your changes.

File: packages/core/src/helpers/calls.ts [hunk 6]
```diff
@@ -20,29 +20,37 @@ export function warnOnInvalidCall(call: Call | Falsy) {
...
+    throw new Error('Missing contract address or method name')
   }
+
   try {
-    const isStatic = queryParams.isStatic ?? queryParams.refresh === 'never'
-    const refreshPerBlocks = typeof queryParams.refresh === 'number' ? queryParams.refresh : undefined
+    contract.interface.encodeFunctionData(method, args)
+    return call
+  } catch (err: any) {
+    throw new Error(`Invalid contract call for method="${method}" on contract="${contract.address}": ${err.message}`)
```
Comment (rzadp) [inline packages/core/src/helpers/calls.ts:36]: Try to include contract name in there, if it exists

Comment (jakvbs) [reply packages/core/src/helpers/calls.ts:33]: I think it's not possible to extract name from Contract instance

**Review by rzadp — APPROVED** [review]

**Pull request merged**